Data/ByteString.hs

```diff
@@ -1419,18 +1422,21 @@ findSubstring pat src
 findSubstrings :: ByteString -- ^ String to search for.
                -> ByteString -- ^ String to seach in.
                -> [Int]
-findSubstrings pat src
-    | null pat        = [0 .. ls]
-    | otherwise       = search 0
+findSubstrings pat = go
   where
-    lp = length pat
-    ls = length src
-    search !n
-        | (n > ls - lp) || null b = []
-        | otherwise = let k = n + length a
-                      in  k : search (k + lp)
+    breakPat = breakSubstring pat
+    go src
+        | null pat        = [0 .. ls]
+        | otherwise       = search 0
       where
-        (a, b) = breakSubstring pat (unsafeDrop n src)
+        lp = length pat
+        ls = length src
```

Comment:
I generally make bindings like these strict to ensure that the compiler doesn't attempt to inline.